### ISSUE TITLE
Improvements to udev rules under Linux

### ISSUE THREAD

**Issue by simon-wh** [WootingKb/wootility-issues#114]
### Pop-up in Wootility
We could have a pop-up on launch on Linux asking about udev rules

e.g.
**Have you installed the udev rules?**
udev rules are required for the keyboard to be detected on Linux

[ Already done it ]
[ Automatic Installation (Requires root access) ]
[ Manual Installation (Guide) ]

### AppImage install script which also does udev

Create install script that is the default download - which pulls latest Wootility build and udev and installs both

*Some internal discussion*
> What would the install script looks like? I remember before we also talked about different package types that can include udev?

An install script would likely look like a file `install.sh` (which you download instead of the AppImage or given as a command to run e.g. https://getantibody.github.io/install/) which you run then asks for sudo permissions to copy of the udev and reload the rules and stuff, then install the AppImage into an appropriate location. Only issue it may have would be it potentially not being the most user-friendly thing ever as in most cases you would need to `chmod +x` it before it can be run. So maybe doing the Wootility modal with auto-install would be a more user-friendly fallback

I would love to do the different Linux package types including udev but given the infinite supply of Linux distros that won't be a solution for all. I defo still want to do a debian package as that will cover a lot of users, but beyond that we need some kind of solution for others

### Additional things to consider

- How to handle updating of udev rules, if we don't go for the approach of a blanket permission-giving on our Vendor ID then we likely would need to update the udev rules as new devices come.

## Approach to the actual udev rules

Currently on [the guide we have](https://wooting.helpscoutdocs.com/article/68-wootility-configuring-device-access-for-wootility-under-linux-udev-rules) it recommends giving permissions to the `input` group and adding yourself to it. There are other approaches like using `TAG+="uaccess"` that gives permissions in a more broader sense, doesn't require any group handling at the cost of giving access more broadly. 

We should evaluate these and determine what approach is best to be the default. 

A couple of other improvements have been brought up:
1. Restricting the permissions to only apply to the interfaces which are actually needed for Wootility and the Analog SDK rather than a blanked unlock of everything
2. As Rocky has suggested now that we have our own Vendor ID we don't necessarily need to specify each individual device, but we should consider if we definitely want to just open up permissions for all devices under our vendor ID

**Comment by Arcitec** [comment]
Wooting's current rules are misguided. Demanding that people add themselves to the `input` group means anything logged in as that user from any source (including SSH) is able to keylog **every** `/dev/input` device without needing root privileges. It's not just annoying to have to add yourself manually to `input`, it's *less secure* to run that way. :smile: 

Let's switch to `uaccess`, which makes systemd automatically grant device access to **only** the currently logged in, local (non SSH) user, to prevent broader keylogging.

**Comment by Arcitec** [comment]
@simon-wh Is this the only place where the rules are hosted, by the way? I haven't been able to find them anywhere on GitHub:

https://help.wooting.io/en/article/wootility-configuring-device-access-for-wootility-under-linux-udev-rules-r6lb2o/

**Comment by Arcitec** [comment]
@simon-wh When switching to the uaccess method, it's important that the rules are in a file that's named something numerically before `73-seat-late.rules` by the way!

https://github.com/systemd/systemd/issues/4288#issuecomment-348166161

**Comment by BigBrainAFK** [comment]
> Wooting's current rules are a bit dangerous.
> 
> Anyone who adds themselves to the `input` group is able to keylog the Wooting keyboard.
> 
> This is what the current Wooting rules allow:
> 
>     * User A is sitting at the keyboard.
> 
>     * User B is logged in via SSH and is in the `input` group.
> 
>     * User B is keylogging everything User A is typing.
> 
> 
> We must add the `uaccess` tag, which makes systemd automatically grant device access to **only** the _currently_ logged in, _local_ (non SSH) user, to prevent broader keylogging.
> 
> It's done by removing your current `MODE` and `GROUP` and replacing them with this:
> 
> ```
> , MODE="0660", TAG+="uaccess"
> ```
> 
> Every file in `/dev/input` is owned by `root:input` and the `MODE` therefore checks that the user is either the owner or part of the same group of the file (`input` group), and the `TAG` ensures that only the physical user has access.
> 
> **Important:** When switching to the uaccess method, it's important that the rules are in a file that's named something numerically before `73-seat-late.rules` by the way!
> 
> [systemd/systemd#4288 (comment)](https://github.com/systemd/systemd/issues/4288#issuecomment-348166161)
> 
> The unofficial standard technique is to use the number `70`. So your rules should be in a file named `70-wooting.rules`.

But if someone gained access to your system you already sort of have bigger issues. the keyboard works without any udev rules but you cant configure it via the app i think.

**Comment by Arcitec** [comment]
> the keyboard works without any udev rules but you cant configure it via the app i think.

I don't own a Wooting keyboard myself, but yes I would assume that typing works without udev rules. Configuration (writing to the `/dev/input` file) requires permissions.

> But if someone gained access to your system you already sort of have bigger issues.

Irrelevant to the discussion. This is about improving Wooting's default udev rules to be safer than they currently are.

The standard is to use `uaccess` tagging:

https://wiki.archlinux.org/title/Udev#Allowing_regular_users_to_use_devices

**Comment by simon-wh** [comment]
@Arcitec Thanks for the detail, I'll get the recommended ones updated based on your recommendations 🙂

**Comment by simon-wh** [comment]
> @simon-wh Is this the only place where the rules are hosted, by the way? I haven't been able to find them anywhere on GitHub:
> 
> https://help.wooting.io/en/article/wootility-configuring-device-access-for-wootility-under-linux-udev-rules-r6lb2o/

Yes, that's currently the only official distribution of the udev rules. 

Also, something to add to the security discussion on things. I'm pretty certain that the Kernel blocks direct reading of the keyboard input hid interfaces regardless of udev rules. Although, we have an interface used for our Analog SDK to get analog values of key presses, so you could vector it through that.

**Comment by Rocky04** [comment]
The tag method should be preferred. I would expect that for HID the standard permissions for a user are enough but that may depends on the configuration and how the distro handle things.

While it will grand access to more interfaces than needed (like the other method) these two rules should be enough for all devices (except the original W1 & W2).

_As far as I know even the mode part isn't needed because if the device node isn't accessed by the user via it's group the default group and user should be root anyway. So I don't get why to even mess with it._

```
SUBSYSTEM=="hidraw", ATTRS{idVendor}=="31e3", TAG+="uaccess"
SUBSYSTEM=="usb", ATTRS{idVendor}=="31e3", TAG+="uaccess"
```

**Comment by Rocky04** [comment]
Please update the whole guide. I feel kinda ashamed that my name is still there.

I would like to see a section for the old method or to have a separate guide for it which is maybe marked as not recommended. 

The text of the current version of the guide don't matches the approach anymore and so is misleading and incorrect. 

Because the tag approach don't need that the user is in the group anymore I don't get why the additional part how the user can be added to the group is still part of the guide.

**Comment by simon-wh** [comment]
I didn't see the updates to this post until now, I've updated the guide now to cut out the stuff with groups and to use the more concise uaccess tag based rules. Looks much cleaner and easier to understand now 👍 

I'd still like to have a flow for this in Wootility itself, so any more feedback on that front would be much appreciated 😁

**Comment by mirh** [comment]
Hope I'm not too much noise, but I wanted to point out that an actual api for analog button has been proposed years ago
https://bugzilla.kernel.org/show_bug.cgi?id=195643#c10
(forgive the nagger that has crashed there today)

**Comment by Rocky04** [comment]
@mirh
That have the analog inputs to do with this thread? This is about udev rules on Linux.

Please open a new feature request and provide a short summary of the linked discussion. Feel free to tag me there.
